rosidl_generator_c/resource/idl__description.c.em

```diff
@@ -0,0 +1,140 @@
+    'detail', convert_camel_case_to_lower_case_underscore(interface_path.stem)]
+include_base = '/'.join(include_parts)
+
+full_type_descriptions = [type_description_msg]
+for service in content.get_elements_of_type(Service):
+  full_type_descriptions.append(extract_subinterface(type_description_msg, 'request_message'))
+  full_type_descriptions.append(extract_subinterface(type_description_msg, 'response_message'))
+  full_type_descriptions.append(extract_subinterface(type_description_msg, 'event_message'))
+for action in content.get_elements_of_type(Action):
+  full_type_descriptions.append(extract_subinterface(type_description_msg, 'goal'))
+  full_type_descriptions.append(extract_subinterface(type_description_msg, 'result'))
+  full_type_descriptions.append(extract_subinterface(type_description_msg, 'feedback'))
+
+  send_goal_service = extract_subinterface(type_description_msg, 'send_goal_service')
+  full_type_descriptions.append(send_goal_service)
+  full_type_descriptions.append(extract_subinterface(send_goal_service, 'request_message'))
+  full_type_descriptions.append(extract_subinterface(send_goal_service, 'response_message'))
+  full_type_descriptions.append(extract_subinterface(send_goal_service, 'event_message'))
+
+  get_result_service = extract_subinterface(type_description_msg, 'get_result_service')
+  full_type_descriptions.append(get_result_service)
+  full_type_descriptions.append(extract_subinterface(get_result_service, 'request_message'))
+  full_type_descriptions.append(extract_subinterface(get_result_service, 'response_message'))
+  full_type_descriptions.append(extract_subinterface(get_result_service, 'event_message'))
+
+  full_type_descriptions.append(extract_subinterface(type_description_msg, 'feedback_message'))
+
+def typename_to_c(typename):
+  return typename.replace('/', '__')
+
+def static_seq(varname, values):
+  """Statically define a runtime Sequence or String type."""
+  if values:
+    return f'{{{varname}, {len(values)}, {len(values)}}}'
+  return '{NULL, 0, 0}'
+}@
+
+#include "@(include_base)__struct.h"
+
+// Declare and define all type names, field names, and default values
+@[for itype_description in all_type_descriptions]@
+@{
+td_c_typename = typename_to_c(itype_description['type_name'])
+}@
+static char @(td_c_typename)__TYPE_NAME[] = "@(itype_description['type_name'])";
+@[  for field in itype_description['fields']]@
+static char @(td_c_typename)__FIELD_NAME__@(field['name'])[] = "@(field['name'])";
+@[    if field['default_value']]@
+static char @(td_c_typename)__DEFAULT_VALUE__@(field['name'])[] = "@(escape_string(field['default_value']))";
+@[    end if]@
+@[  end for]@
+
+@[end for]@
+@
+/// Define all arrays of Fields
+@[for itype_description in all_type_descriptions]@
+@{
+td_c_typename = typename_to_c(itype_description['type_name'])
+}@
+
+@[  if itype_description['fields']]@
+static rosidl_runtime_c__type_description__Field @(td_c_typename)__FIELDS[] = {
```

Comment:
Should this be `static const`?

Reply:
It can't, because `rosidl_runtime_c__type_description__Field__Sequence` is defined as

```
typedef struct rosidl_runtime_c__type_description__Field__Sequence
{
  rosidl_runtime_c__type_description__Field * data;
  /// The number of valid items in data
  size_t size;
  /// The number of allocated items in data
  size_t capacity;
} rosidl_runtime_c__type_description__Field__Sequence;
```

So a const ptr couldn't be used in a brace  initializer for that type. Same goes for `rosidl_runtime_c__String`, you may have not noticed the `static char foo[] = "bar";` for type names, field names, default values.

Reply:
It turns out that Windows won't allow for the final `rosidl_runtime_c__type_description__TypeDescription` definitions with the array pointers though, so I might need a different tactic like inlining some initializers via empy or `#define`s.

I really wanted to avoid having to use a getter-function, but maybe that's the way that it has to go to initialize some values of the structs.